src/python/pants/option/arg_splitter_test.py

```diff
@@ -54,37 +54,60 @@ def assert_valid_split(
     splitter = ArgSplitter(ArgSplitterTest._known_scope_infos)
     args = shlex.split(args_str)
     split_args = splitter.split_args(args)
-    self.assertEqual(expected_goals, split_args.goals)
-    self.assertEqual(expected_scope_to_flags, split_args.scope_to_flags)
-    self.assertEqual(expected_specs, split_args.specs)
-    self.assertEqual(expected_passthru, split_args.passthru)
-    self.assertEqual(expected_passthru_owner, split_args.passthru_owner)
-    self.assertEqual(expected_is_help, splitter.help_request is not None)
-    self.assertEqual(expected_help_advanced,
-                      (isinstance(splitter.help_request, OptionsHelp) and
-                       splitter.help_request.advanced))
-    self.assertEqual(expected_help_all,
-                      (isinstance(splitter.help_request, OptionsHelp) and
-                       splitter.help_request.all_scopes))
-    self.assertEqual(expected_unknown_scopes, split_args.unknown_scopes)
+    assert expected_goals == split_args.goals
+    assert expected_scope_to_flags == split_args.scope_to_flags
+    assert expected_specs == split_args.specs
+    assert expected_passthru == split_args.passthru
+    assert expected_passthru_owner == split_args.passthru_owner
+    assert expected_is_help == (splitter.help_request is not None)
+    assert expected_help_advanced == (
+      isinstance(splitter.help_request, OptionsHelp) and splitter.help_request.advanced
+    )
+    assert expected_help_all == (
+      isinstance(splitter.help_request, OptionsHelp) and splitter.help_request.all_scopes
+    )
+    assert expected_unknown_scopes == split_args.unknown_scopes
+
+  def assert_unknown_goal(self, args_str: str, unknown_goals: List[str]) -> None:
+    splitter = ArgSplitter(ArgSplitterTest._known_scope_infos)
+    result = splitter.split_args(shlex.split(args_str))
+    assert isinstance(splitter.help_request, UnknownGoalHelp)
+    assert set(unknown_goals) == set(splitter.help_request.unknown_goals)
+    assert result.unknown_scopes == unknown_goals
 
   def test_is_spec(self) -> None:
     def assert_spec(arg: str) -> None:
-      self.assertTrue(ArgSplitter.spec(arg))
+      assert ArgSplitter.spec(arg) is True
 
     def assert_not_spec(arg: str) -> None:
-      self.assertFalse(ArgSplitter.spec(arg))
+      assert ArgSplitter.spec(arg) is False
 
-    assert_spec('a/b/c')
-    assert_spec('a/b:c')
-    assert_spec(':c')
-    with temporary_dir() as tmpdir:
-      os.mkdir(os.path.join(tmpdir, 'foo'))
-      with pushd(tmpdir):
-        assert_spec('foo')
+    unambiguous_specs = [
+      'a/b/c',
+      'a/b/c/',
+      'a/b:c',
+      'a/b/c.txt',
+      ':c',
+      '::',
+      'a/',
+      './a.txt',
+      '.',
```

Comment:
I don't know how I feel about this one. What are the semantics of `./pants cloc .`? I'd rather that we don't allow this, at least for now.

If you all agree, we need to decide whether the validation should happen this early on or later when trying to expand a raw spec into a `FilesystemSpec`? If we do it this early on, then `./pants cloc .` would complain that `.` is an unknown goal. I think I lean towards that approach.

Reply:
How would you "run cloc over everything under the buildroot"? `./pants cloc .` seems like a sensible way to do this.

Reply:
Hm, we actually have logic already for normalizing `.`. We might need to revisit it, but I agree with your proposed semantics.

https://github.com/pantsbuild/pants/blob/261849bdd9032101ad5e76519c30a2b4d0784b08/src/python/pants/base/cmd_line_spec_parser.py#L47-L48

```
▶ ./pants list .
Scrubbed PYTHONPATH=/Users/eric/DocsLocal/code/projects/pants/src/python: from the environment.
timestamp: 2020-01-09T22:54:04.105785
Exception caught: (pants.engine.scheduler.ExecutionError)
  File "/Users/eric/DocsLocal/code/projects/pants/src/python/pants/bin/pants_loader.py", line 85, in <module>
    main()
  File "/Users/eric/DocsLocal/code/projects/pants/src/python/pants/bin/pants_loader.py", line 81, in main
    PantsLoader.run()
  File "/Users/eric/DocsLocal/code/projects/pants/src/python/pants/bin/pants_loader.py", line 77, in run
    cls.load_and_execute(entrypoint)
  File "/Users/eric/DocsLocal/code/projects/pants/src/python/pants/bin/pants_loader.py", line 70, in load_and_execute
    entrypoint_main()
  File "/Users/eric/DocsLocal/code/projects/pants/src/python/pants/bin/pants_exe.py", line 32, in main
    PantsRunner(start_time=start_time).run()
  File "/Users/eric/DocsLocal/code/projects/pants/src/python/pants/bin/pants_runner.py", line 106, in run
    return runner.run()
  File "/Users/eric/DocsLocal/code/projects/pants/src/python/pants/bin/local_pants_runner.py", line 253, in run
    self._run()
  File "/Users/eric/DocsLocal/code/projects/pants/src/python/pants/bin/local_pants_runner.py", line 331, in _run
    engine_result = self._maybe_run_v2()
  File "/Users/eric/DocsLocal/code/projects/pants/src/python/pants/bin/local_pants_runner.py", line 301, in _maybe_run_v2
    self._target_roots,
  File "/Users/eric/DocsLocal/code/projects/pants/src/python/pants/init/engine_initializer.py", line 225, in run_goal_rules
    exit_code = self.scheduler_session.run_goal_rule(goal_product, params)
  File "/Users/eric/DocsLocal/code/projects/pants/src/python/pants/engine/scheduler.py", line 471, in run_goal_rule
    self._trace_on_error([exc], request)
  File "/Users/eric/DocsLocal/code/projects/pants/src/python/pants/engine/scheduler.py", line 449, in _trace_on_error
    unique_exceptions,

Exception message: 1 Exception encountered:
Computing Select((<pants.engine.console.Console object at 0x1b91af39668>+AddressSpecs(dependencies=(SingleAddress(directory='', name=''),), matcher=AddressSpecsMatcher(tags=(), exclude_patterns=()))+OptionsBootstrapper(env_tuples=(('PANTS_DEV', '1'),), bootstrap_args=(), args=('/Users/eric/DocsLocal/code/projects/pants/src/python/pants/bin/pants_loader.py', 'list', '.'), config=_ChainedConfig(chained_configs=(<pants.option.config._SingleFileConfig object at 0x10d022f28>,)))), List)
  Computing Task(list_targets(), (<pants.engine.console.Console object at 0x1b91af39668>+AddressSpecs(dependencies=(SingleAddress(directory='', name=''),), matcher=AddressSpecsMatcher(tags=(), exclude_patterns=()))+OptionsBootstrapper(env_tuples=(('PANTS_DEV', '1'),), bootstrap_args=(), args=('/Users/eric/DocsLocal/code/projects/pants/src/python/pants/bin/pants_loader.py', 'list', '.'), config=_ChainedConfig(chained_configs=(<pants.option.config._SingleFileConfig object at 0x10d022f28>,)))), List, false)
    Computing Task(remove_provenance(), AddressSpecs(dependencies=(SingleAddress(directory='', name=''),), matcher=AddressSpecsMatcher(tags=(), exclude_patterns=())), BuildFileAddresses, true)
      Computing Task(provenanced_addresses_from_address_families(), AddressSpecs(dependencies=(SingleAddress(directory='', name=''),), matcher=AddressSpecsMatcher(tags=(), exclude_patterns=())), ProvenancedBuildFileAddresses, true)
        Throw("" was not found in namespace "". Did you mean one of:
  :3rdparty_directory
  :avro-java-runtime
  :build_root
  :build_tools
  :gitignore
  :isort_cfg
  :javac9
  :kythe-java-extractor
  :kythe-java-indexer
  :pants_ini
  :pants_pex
  :pyproject
  :scala-js-cli
  :scala-js-compiler
  :scala-js-library
  :scalajs_3rdparty_directory
  :scrooge-gen
  :scrooge-linter)
          Traceback (most recent call last):
            File "/Users/eric/DocsLocal/code/projects/pants/src/python/pants/engine/build_files.py", line 231, in provenanced_addresses_from_address_families
              addr_families_for_spec
            File "/Users/eric/DocsLocal/code/projects/pants/src/python/pants/base/specs.py", line 132, in address_target_pairs_from_address_families
              raise self._SingleAddressResolutionError(single_af, self.name)
          pants.base.specs.SingleAddress._SingleAddressResolutionError

          The above exception was the direct cause of the following exception:

          Traceback (most recent call last):
            File "/Users/eric/DocsLocal/code/projects/pants/src/python/pants/engine/native.py", line 486, in extern_generator_send
              res = c.from_value(func[0]).send(c.from_value(arg[0]))
            File "/Users/eric/DocsLocal/code/projects/pants/src/python/pants/engine/build_files.py", line 239, in provenanced_addresses_from_address_families
              _raise_did_you_mean(e.single_address_family, e.name, source=e)
            File "/Users/eric/DocsLocal/code/projects/pants/src/python/pants/engine/build_files.py", line 75, in _raise_did_you_mean
              raise resolve_error from source
          pants.engine.mapper.ResolveError: "" was not found in namespace "". Did you mean one of:
            :3rdparty_directory
            :avro-java-runtime
            :build_root
            :build_tools
            :gitignore
            :isort_cfg
            :javac9
            :kythe-java-extractor
            :kythe-java-indexer
            :pants_ini
            :pants_pex
            :pyproject
            :scala-js-cli
            :scala-js-compiler
            :scala-js-library
            :scalajs_3rdparty_directory
            :scrooge-gen
            :scrooge-linter
```